### ISSUE TITLE
Add TCP/HTTP/network requests to the CRI API

### ISSUE THREAD

**Issue by aojea** [kubernetes/kubernetes#102613]
#### What would you like to be added:

CRI API methods to execute kubelet TCP and HTTP probes/lifecycle hooks

#### Why is this needed:

Current probes/lifecycle hooks implementation forces the networking model to provide connectivity between the kubelet and the pods, however, this may not be true or easy to achieve in all platforms and all the CRI implementations.

Since the CRI API abstracts the communication between "the kubelet and its pods", and it already implements the exec probes through this API, it may be useful to have an API call to exec the networking TCP and/or HTTP probes.

This request doesn't require current implementations to move to a model based on CRI/API calls, it just opens the possibility to new platforms and other CRIs to start using it, and the project to study if it should start to migrate to this mode.

However, this can present some challenges as per Tim's comment

> While this fixes some issues, it introduces new ones - old probes that used the node's localhost will be broken, for example.  I don't see how we could do it by default.

References:
- sig-network thread https://groups.google.com/g/kubernetes-sig-network/c/nIo6ffat7TI/m/xIEtpAAeDgAJ
- Pod readiness probe cannot be directed at specific IP family
https://github.com/kubernetes/kubernetes/issues/101324#issuecomment-831563048
- Pod probes lead to blind SSRF from the node
https://github.com/kubernetes/kubernetes/issues/99425#issuecomment-829712715

**Comment by thockin** [comment]
Added sig-node.  Not sure who in particular feels ownership of CRI these day?  @dchen1107 

One of the main goals here is to close off a network-policy bypass vector, using the "host" field of probes.   A side-benefit is that probes could bind() to localhost only.  Are there other big reasons, @aojea ?

I don't have creative thought about how to fix cases that probe against node-localhost today.  We'd have to special case that or something, which is ugly.

**Comment by aojea** [comment]
> . Are there other big reasons, @aojea ?

It should solve dual stack probes too, you can use localhost and the os will pick the right family

**Comment by thockin** [comment]
Can you explain?  Today localhost means the node.

**Comment by aojea** [comment]
> Can you explain? Today localhost means the node.

Yeah, and that causes a possible security problem as per `Pod probes lead to blind SSRF from the node ` #99425

> Other possible targets are localhost endpoints, such as 127.0.0.1:2381/metrics (etcd metrics), or external targets (httpbin.org, for example).

The other issue we have is `Pod readiness probe cannot be directed at specific IP family` #101324.
Currently,  if you don't specify the `Host` field in the probe, the probe Status.PodIPs[0] , however, in a dual stack cluster you can have a single stack application that only listen in Status.PodIPs[1].

One of the solutions suggested is to implement in the probe logic a fallback mechanism: "try Status.PodIPs[0] and if it fails I try Status.PodIPs[1]" mimicking the happy-eyeballs algorithm.

If we can execute the probe inside the namespace and if we use `localhost`,  the OS will (assume normal deployments) return ::1 and 127.0.0.1 and golang net/http, curl , ... already implements happy eyeballs. 

Actually the solution to the problem I suggested is to use an exec probe and curl https://github.com/kubernetes/kubernetes/issues/101324#issuecomment-829910810

```
curl localhost -v
* Rebuilt URL to: localhost/
*   Trying ::1...
* TCP_NODELAY set
* connect to ::1 port 80 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* connect to 127.0.0.1 port 80 failed: Connection refused
* Failed to connect to localhost port 80: Connection refused
* Closing connection 0
````

In addition, if we remove all the network probe logic from kubelet and offload to the runtimes, I wonder if this will be fixed https://github.com/kubernetes/kubernetes/issues/89898#issuecomment-817675889 too :thinking:

**Comment by thockin** [comment]
Got it.  The problem is compat - there are probes that hit "localhost"
today and if we change the meaning of that from "node" to "pod", they
break.  I think that, at the very least, we'd need to special-case
"localhost" to mean the node's IP.

**Comment by khenidak** [comment]
@aojea  does all CRI implement that api?

**Comment by aojea** [comment]
> @aojea does all CRI implement that api?

That's the proposal, introduce the new API and work with runtimes to implement it

**Comment by danwinship** [comment]
> Since the CRI API abstracts the communication between "the kubelet and its pods", and it already implements the exec probes through this API, it may be useful to have an API call to exec the networking TCP and/or HTTP probes.

You never actually say that you mean "having the CRI implement the probes _in the pod's network namespace_". (You could add a CRI API for TCP/HTTP probes, but have the runtime implement it in exactly the same way kubelet does.)

> I don't have creative thought about how to fix cases that probe against node-localhost today.

What is the use case for a non-hostNetwork pod to probe against node-localhost? I guess maybe if it has a HostPort, but we could detect that case. And maybe if it has a NodePort, which would be a lot harder...

**Comment by mikebrow** [comment]
SGTM to add a CRI Probe() API, when can we start.

**Comment by aojea** [comment]
> SGTM to add a CRI Probe() API, when can we start.

at this point I really don't have time, but the next step should be to create a KEP and get it approved for 1.25

**Comment by aojea** [comment]
/reopen
/remove-lifecycle rotten

This is important for:
- kubelet performance and failing probes because of that
- kubelet security, to remove the requirement to access any other pod in the cluster
- multi-network, as a consequence of the previous one

**Comment by danwinship** [comment]
So one way to deal with the compatibility issues would be to deprecate the `httpGet` and `tcpSocket` probe types and replace them with new versions (`http` and `tcp`?) that have a nearly-identical kube API (ie, most users can just "`s/httpGet/http/; s/tcpSocket/tcp/`" on their pods and everything will just work), but are implemented by kubelet using the new CRI API. (Not sure about the `grpc` type... it might be OK? And it's still beta so even if it needs changes maybe we can just change it?)

Then (as @tallclair suggested in one of the linked issues) we can use PodSecurity admission to let administrators ban the `httpGet` and `tcpSocket` versions to force their users to upgrade to the new, safer versions.

I don't think we can compatibly drop the NetworkPolicy "all pods allow ingress from their node IP" rule even if we get everyone to uprade their probes. (There may be pods that assume that access is allowed from their node IP for things other than probes.) But assuming "NPv2" happens, we can drop the special rule then; if people want to use deprecated probe types on a pod protected by NPv2, then they need to manually add a policy allowing the probes.

**Comment by uablrek** [comment]
There is a _very serious_ performance problem with exec'ing into a POD described for instance in:

https://medium.com/netcracker/exec-probes-a-story-about-experiment-and-relevant-problems-12de616c0a76

I think this is relevant whenever the exec is user initiated (with an exec probe) or done by the CRI-plugin.

**Comment by aojea** [comment]
hmm, is like http doesn't do syscall to the kernel too?

also, runc had several regressions in performance during that time, I wonder if is a fair comparison, I would love to have a more exhaustive analysis of the  performance of each probe before claiming that exec probes are more expensive that http ones

**Comment by uablrek** [comment]
We have had performance problems with exec probes. Unfortunately I was not involved, but I'll try to get some info. I think background load went from ~30% with exec probes to ~3% with http probes, but as said I was not involved.

**Comment by uablrek** [comment]
I am not saying that exec probes can't be used, but that the performance problem _must_ be addressed.

**Comment by aojea** [comment]
> I am not saying that exec probes can't be used, but that the performance problem _must_ be addressed.

we are arguing that it is not clear those claims in the environment are generalizable

 Your point is valid Lars, if this move forward we should analyze if there is a performance regression, that is for the KEP

However, this is hard to make it backward compatible :/

https://github.com/kubernetes/kubernetes/issues/102613#issuecomment-856889440
